### ISSUE TITLE
Broken link in repo README.

### ISSUE THREAD

**Issue by VorpalBlade** [rustls/rustls#1275]
The link to supported platforms point to https://github.com/briansmith/ring#online-automated-testing

There is no such section in the ring readme, the link is broken.

**Comment by cpu** [comment]
Good catch, thanks for opening an issue :+1: 

It looks like the upstream README dropped this content in https://github.com/briansmith/ring/commit/2047b68fa2aab521a57a445b0398760032226362. I don't see it having been migrated anywhere we can link to. Perhaps we should enumerate the supported platforms ourselves and keep it in-sync with upstream?

**Comment by djc** [comment]
> It looks like the upstream README dropped this content in [briansmith/ring@2047b68](https://github.com/briansmith/ring/commit/2047b68fa2aab521a57a445b0398760032226362). I don't see it having been migrated anywhere we can link to. Perhaps we should enumerate the supported platforms ourselves and keep it in-sync with upstream?

Hmm, seems a bit of a pain in the ass, especially as we bring pluggable crypto providers in.

**Comment by cpu** [comment]
> Hmm, seems a bit of a pain in the ass, especially as we bring pluggable crypto providers in.

Fair point.

Looking at the README again I think we already have enough information there to drop the link without adding more detail. It says:
>  At the time of writing this means x86, x86-64, armv7, and aarch64.

I'll put up a PR to remove the dead Ring link and we can try to keep this bit of text up to date. I don't think it will be a big burden since Ring development is pretty slow paced, pluggable crypto providers don't exist yet, and new arch target support doesn't happen with much frequency.

**Comment by VorpalBlade** [comment]
I don't think that quite answers the question though. When it says "ARMv7", does it mean with hard float or not? What about musl vs glibc? What about ARM on Windows? Does any of that even matter?

Specific target triplets (or target triplet globs (e.g. `armv7-unknown-linux-*`) would be more useful.

**Comment by VorpalBlade** [comment]
> The whole mess of trying to explain ARMv7 is exactly the kind of thing that we should probably avoid, since this project is not the limiting factor...

As a potential user who have to decide between using rustls or OpenSSL bindings for my project this is however very useful information. Both have downsides: OpenSSL supports most platforms and operating systems, but involves relying on the platform to provide it (or vendoring it., either statically or dynamically linked). Rustls has less platform support, but everything is linked statically and standalone.

To make an informed decision I need those complicated details.

**Comment by djc** [comment]
I completely agree that you need those details. I'm just saying that the rustls implementation itself is in no way platform-dependent, so it feels a little awkward to document these constraints here, when really there should be some part of *ring*'s documentation that we can point to. Unfortunately *ring* isn't getting a lot of maintenance right now...

**Comment by ctz** [comment]
Can we instead link to ring's https://github.com/briansmith/ring/blob/main/.github/workflows/ci.yml or some derivative? That precisely documents which platforms it supports and is tested on.

**Comment by djc** [comment]
Good idea. I think the relevant list of targets is here: https://github.com/briansmith/ring/blob/b/0.16/.github/workflows/ci.yml#L152.

**Comment by VorpalBlade** [comment]
> Good idea. I think the relevant list of targets is here: https://github.com/briansmith/ring/blob/b/0.16/.github/workflows/ci.yml#L152.

And that is exactly what I needed. Which shows that while two ARMv7 targets (as well as several other ARM targets) are supported, armv7-unknown-linux-gnueabihf (to target Raspberry Pi) is not one of them, and this is one I absolutely need. Even though I suspect it might work (based on supporting very similar targets, including the musl version), that is not guaranteed. So OpenSSL it is for me.

**Comment by jsha** [comment]
@VorpalBlade do you have access to a Raspberry Pi to try building and `cargo test`ing *ring* on? I would be curious to know if it works or how it breaks.

Note that even armv7-unknown-linux-musleabihf, which is on the CI list, is run with `cargo test --no-run`, in other words it only builds the tests but does not run them. This is because GitHub Actions does not have any runners of the appropriate type. To fully support various ARM platforms (including running tests) would require appropriate runners (which could be third-party runners).

**Comment by VorpalBlade** [comment]
@jsha I use [cross](https://github.com/cross-rs/cross) for one of my projects to run tests on ARM. It works using qemu-user. So while it is not the real thing, it does at least execute the ARM code emulated. And it seems to work fine on GitHub Actions

As it happens I *do* have access to both a Pi 2 and a Pi 4 as well. I will look into running those tests tomorrow.

**Comment by jsha** [comment]
Thanks for the tip! That was really easy. I used `cross` to run tests for *ring* on armv7-unknown-linux-gnueabihf and they passed. 🎉. I'm running them for rustls now, but I expect they'll pass. Do you have any interest in contributing a PR to test armv7-unknown-linux-gnueabihf via `cross` as part of rustls' and *ring*'s respective CI? That would allow us to more confidently call the platform fully supported, which I think would meet your needs, right?

**Comment by VorpalBlade** [comment]
I would solve my concern. However this is for a personal hobby project. I cannot guarantee when I will get time for it. Plus I really really hate dealing with CI configs. It is likely a lot of people would be better at doing this than me.
